bundle/config/mutator/python/override.go

```diff
@@ -0,0 +1,313 @@
+package python
+
+import (
+	"fmt"
+
+	"github.com/databricks/cli/libs/dyn"
+	"github.com/databricks/cli/libs/dyn/merge"
+)
+
+// overrideMode restricts which changes are allowed during the override process
+type overrideMode string
+
+const OverrideModeAppend overrideMode = "append"
+const OverrideModeAppendOrModify overrideMode = "append-or-update"
+
+// isModified keeps track of whether a value has been modified during the override process
+type isModified struct {
+	value bool
+}
+
+var Modified = isModified{value: true}
+var NotModified = isModified{value: false}
+
+type overrideFunc func(dyn.Value, dyn.Value) (dyn.Value, isModified, error)
+
+// OverrideByPath overrides bundle config 'leftRoot' with 'rightRoot' values in 'path', keeping
+// 'location' if values haven't changed. This allows to keep track of where the values
+// were last modified for error reporting.
+//
+// 'path' must point to a mapping or not exist, otherwise, error is returned
+//
+// 'mode' allows to restrict what kinds of changes are allowed in 'path':
+// - append: only new values can be added to mapping
+// - append-or-modify: new values can be added, or existing values can be modified
+func OverrideByPath(leftRoot dyn.Value, rightRoot dyn.Value, path dyn.Path, mode overrideMode) (dyn.Value, error) {
+	_, leftErr := dyn.GetByPath(leftRoot, path)
+	_, rightErr := dyn.GetByPath(rightRoot, path)
+
+	if leftErr != nil && rightErr != nil {
+		return leftRoot, nil
+	}
+
+	leftRoot, err := insertMappingIfAbsent(leftRoot, path)
+
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	rightRoot, err = insertMappingIfAbsent(rightRoot, path)
+
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	return dyn.Map(leftRoot, path.String(), func(p dyn.Path, left dyn.Value) (dyn.Value, error) {
+		leftMapping := left.MustMap()
+		rightMapping, err := dyn.GetByPath(rightRoot, p)
+
+		if err != nil {
+			return dyn.InvalidValue, err
+		}
+
+		out, _, err := overrideMapping(
+			leftMapping,
+			rightMapping.MustMap(),
+			func(left dyn.Value, right dyn.Value) (dyn.Value, isModified, error) {
+				merged, modified, err := override(left, right)
+
+				if err != nil {
+					return dyn.InvalidValue, NotModified, err
+				}
+
+				if modified == Modified {
+					if mode == OverrideModeAppend && left != dyn.NilValue {
+						return dyn.InvalidValue, NotModified, unexpectedChangeError(left.Location())
+					}
+
+					if mode == OverrideModeAppendOrModify && right == dyn.NilValue {
+						return dyn.InvalidValue, NotModified, unexpectedChangeError(left.Location())
+					}
+				}
+
+				return merged, modified, nil
```

Comment:
This hook is executed only for the top level value, not for inner values. Is this intentional?

Reply:
Yes, because we want to handle top-level value from inner values differently. E.g., appending to top-level is allowed during pre-init, but appending to nested field is not allowed, because it will mutate existing resource.